core/server/proxy/src/main/java/alluxio/proxy/s3/S3Handler.java

```diff
@@ -0,0 +1,518 @@
+  private static final Logger LOG = LoggerFactory.getLogger(S3Handler.class);
+  private static final ThreadLocal<byte[]> TLS_BYTES =
+          ThreadLocal.withInitial(() -> new byte[8 * 1024]);
+  private final String mBucket;
+  private final String mObject;
+  private final HttpServletRequest mServletRequest;
+  private final HttpServletResponse mServletResponse;
+  private final ServletContext mServletContext;
+  public AsyncUserAccessAuditLogWriter mAsyncAuditLogWriter;
+  String[] mUnsupportedSubResources = {"acl", "policy", "versioning", "cors",
+    "encryption", "intelligent-tiering", "inventory", "lifecycle",
+    "metrics", "ownershipControls", "replication", "website", "accelerate",
+    "location", "logging", "metrics", "notification", "ownershipControls",
+    "policyStatus", "requestPayment", "attributes", "legal-hold", "object-lock",
+    "retention", "torrent", "publicAccessBlock", "restore", "select"};
+  Set<String> mUnsupportedSubResourcesSet = new HashSet<>(Arrays.asList(mUnsupportedSubResources));
+  Map<String, String> mAmzHeaderMap = new HashMap<>();
+  Request mBaseRequest;
+  private Stopwatch mStopwatch;
+  private String mUser;
+  private S3BaseTask mS3Task;
+  private FileSystem mMetaFS;
+
+  /**
+   * S3Handler Constructor.
+   * @param bucket
+   * @param object
+   * @param request
+   * @param response
+   */
+  public S3Handler(String bucket, String object,
+                   HttpServletRequest request, HttpServletResponse response) {
+    mBucket = bucket;
+    mObject = object;
+    mServletRequest = request;
+    mServletResponse = response;
+    mServletContext = request.getServletContext();
+  }
+
+  /**
+   * Create a S3Handler based on the incoming Request.
+   * @param path
+   * @param request
+   * @param response
+   * @return A S3Handler
+   * @throws Exception
+   *
+   */
+  public static S3Handler createHandler(String path,
+                                        HttpServletRequest request,
+                                        HttpServletResponse response) throws Exception {
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    Matcher bucketMatcher = BUCKET_PATH_PATTERN.matcher(path);
+    Matcher objectMatcher = OBJECT_PATH_PATTERN.matcher(path);
+
+    String pathStr = path;
+    String bucket = null;
+    String object = null;
+    S3Handler handler = null;
+    try {
+      if (bucketMatcher.matches()) {
+        pathStr = path.substring(S3RequestServlet.S3_V2_SERVICE_PATH_PREFIX.length() + 1);
+        bucket = URLDecoder.decode(pathStr, "UTF-8");
+      } else if (objectMatcher.matches()) {
+        pathStr = path.substring(S3RequestServlet.S3_V2_SERVICE_PATH_PREFIX.length() + 1);
+        bucket = URLDecoder.decode(
+          pathStr.substring(0, pathStr.indexOf(AlluxioURI.SEPARATOR)), "UTF-8");
+        object = URLDecoder.decode(
+          pathStr.substring(pathStr.indexOf(AlluxioURI.SEPARATOR) + 1), "UTF-8");
```

Comment:
If the path is `api/v1/s3/bucket/`, it will enter here and the bucket is `bucket`, the object is `/`?

Reply:
sry, my misunderstanding, the object will be "".

Reply:
yes, object will empty str : )